### PR TITLE
Improve docs and tests for `zircon_object::signal`

### PR DESCRIPTION
**Pull request by xxchan**
# Content of this PR
- `zircon_object::signal` passed `#![deny(missing_docs)]`, but `#![allow(missing_docs)]` is added in `signal::port_packet`.
Since `port_packet` just defines some C-style data types and the names can convey meanings, I think it is fine to omit docs.
- `zircon_object::signal` test coverage increases to 96.4%.
- Some minor refactors
  - validate `options` in `Port::new()`
  - refactor `FutexFuture::drop`: using `waiter.inner.futex` instead of `waiter.thread.proc().get_futex()`
- Some minor fixes
  - The doc of `port.wait` says it takes out all packets, but only takes out one in fact.
  - The return count of `futex.wake` is wrong.

PTAL and see whether the changes are proper. Besides, I still have
# Some questions
## timer
1. `Slack` is useless. What's the plan about it?

## port
1. Is it OK that many structs in `port_packet` are public?
2. Where should the validation of `options` be done? Low level `Port::new()` or high level syscalls? Or both? Should `Interrupt::bind()` validate `port.can_bind_to_interrupt()` again? (It is not validated now, so we can pass a port without `BIND_TO_INTERUPT` option actually)
I'm a little bit confused about such kind of questions. 

## futex
**I'm not very familiar with futex, so I might have understood some concepts wrong**

1. Should futex only be created by `proc.get_futex()` and not by `Futex::new`? Otherwise the futex cannot be bound to some thread. This question arises from this situation:
```rust
static VALUE: AtomicI32 = AtomicI32::new(1);
let futex = Futex::new(&VALUE);

let root_job = Job::root();
let proc = Process::create(&root_job, "proc", 0).expect("failed to create process");
let thread = Thread::create(&proc, "thread", 0).expect("failed to create thread");
assert!(!Arc::ptr_eq(&futex, &thread.proc().get_futex(&VALUE)));
```
This is kind of like the question mentioned above about validating port `options`, where the low level primitives are misused . So who is responsible to ensure the correctness? Should the object methods be robust enough themselves, or just let the user obey some contracts?
2. What's the use of `waiter.thread` (and the parameter `thread` in `Futex::wait_with_owner()`)? It is reasonable that the waiter knows which thread is waiting...But it can be set to `None`? And I think it is not used anywhere actually.

### CODE REVIEW

**Comment by xxchan** [comment]
One more question: I found [async-booc says here](https://rust-lang.github.io/async-book/02_execution/03_wakeups.html) that:
> Importantly, we have to update the `Waker` every time the future is polled because the future may have moved to a different task with a different `Waker`. This will happen when futures are passed around between tasks after being polled.

Is that needed in zCore?

**Comment by xxchan** [comment]
Another question: there is a `EventPair::peer()`, `event0.peer()` will call it instead of the trait function. Is this intended?

**Review by wangrunji0408 — APPROVED** [review]
Looks good!

**Comment by wangrunji0408** [comment]
Sorry for my late response. I was busy on the zCore summer of code last week. QAQ

File: zircon-object/src/signal/futex.rs [hunk 1]
```diff
@@ -321,5 +329,97 @@ mod tests {
...
+            let futex = futex.clone();
+            async_std::task::spawn(async move {
+                async_std::task::sleep(Duration::from_millis(10)).await;
+                VALUE.store(2, Ordering::SeqCst);
+                // inconsistent value should fail.
+                assert_eq!(
+                    futex.requeue(1, 1, 1, &requeue_futex, None),
+                    Err(ZxError::BAD_STATE)
+                );
+                assert!(futex.requeue(2, 1, 1, &requeue_futex, None).is_ok());
```
[the remark this answers is not in the document]
Comment (xxchan) [reply zircon-object/src/signal/futex.rs:370]: done

**Comment by wangrunji0408** [comment]
## About the Questions

### Timer
> Slack is useless. What's the plan about it?

Seems it's not critical. We can consider removing this code.

### Port
> Is it OK that many structs in port_packet are public?

I prefer to make them private if possible.

> Where should the validation of options be done? Low level Port::new() or high level syscalls? Or both?

High level syscalls, I think.

> Should Interrupt::bind() validate port.can_bind_to_interrupt() again? (It is not validated now, so we can pass a port without BIND_TO_INTERUPT option actually)

Yes. So you found a bug :)

### Futex

> Should futex only be created by proc.get_futex() and not by Futex::new?

For syscall layer, use `proc.get_futex()` only. This is a kind of contract but never be mentioned before...

> So who is responsible to ensure the correctness? Should the object methods be robust enough themselves, or just let the user obey some contracts?

In general, objects should be designed to be robust enough. But in the case of futex, it doesn't know the concept of memory space or process. So when using futex within multiple processes, it's the user(syscall layer)'s response to make it correct.

> What's the use of waiter.thread (and the parameter thread in Futex::wait_with_owner())?

I'm not familiar with the implementation details of futex with threads. Maybe @PanQL and @benpigchu can give a help.

### Waker

> Importantly, we have to update the Waker every time the future is polled because the future may have moved to a different task with a different Waker.

Yes I noticed that. This rule should be followed in all leaf future implementations. Although this issue has not brought trouble so far, it should be fixed later.

### EventPair

> there is a EventPair::peer(), event0.peer() will call it instead of the trait function. Is this intended?

Yes. When you get an object which is known to be a `EventPair`, its peer should be the same type intuitively.

**Comment by xxchan** [comment]
Thank you for answering me ☺

**Comment by xxchan** [comment]
> >Where should the validation of options be done? Low level Port::new() or high level syscalls? Or both?
>
> High level syscalls, I think.

Is that because we can early exit the syscall if the option is invalid?

If so, we can use `from_bits()`  in syscall, and `from_bits_truncate()`  in object.

**Comment by xxchan** [comment]
Oh, we can directly use `PortOptions` in objects

**Pull request merged**